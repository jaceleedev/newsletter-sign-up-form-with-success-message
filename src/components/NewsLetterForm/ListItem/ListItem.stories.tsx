import type { Meta, StoryObj } from '@storybook/react';

import ListItem from './ListItem';

const meta = {
  component: ListItem,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    content: {
      description: 'Text to be displayed in the ListItem',
      control: 'text',
    },
  },
  decorators: [
    (Story) => (
      <ul
        style={{
          display: 'flex',
          flexDirection: 'column',
          gap: '10px',
          width: '327px',
        }}
      >
        <Story />
      </ul>
    ),
  ],
} satisfies Meta<typeof ListItem>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    content: 'Product discovery and building what matters',
  },
};

export const MultipleListItems: Story = {
  render: () => (
    <>
      <ListItem content="Product discovery and building what matters" />
      <ListItem content="Measuring to ensure updates are a success" />
      <ListItem content="And much more!" />
    </>
  ),
  args: {
    content: '',
  },
};
