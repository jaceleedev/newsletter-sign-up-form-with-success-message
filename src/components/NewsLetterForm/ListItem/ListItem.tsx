import Image from 'next/image';

export interface ListItemProps {
  content: string;
}

function ListItem({ content }: Readonly<ListItemProps>) {
  return (
    <li className="grid grid-cols-[21px_1fr] gap-4 w-full">
      <Image
        src="/images/icon-list.svg"
        alt=""
        width={21}
        height={21}
        aria-hidden={true}
        priority
        className="mt-[2px]"
      />
      <p className="text-body text-dark-navy">{content}</p>
    </li>
  );
}

export default ListItem;
