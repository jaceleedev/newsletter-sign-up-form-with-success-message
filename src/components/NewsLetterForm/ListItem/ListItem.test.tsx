import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import ListItem from './ListItem';

expect.extend(toHaveNoViolations);

describe('ListItem', () => {
  // 1. Rendering test
  it('renders correctly', () => {
    render(<ListItem content="Render test" />);
    const listItem = screen.getByText('Render test').closest('li');
    expect(listItem).toBeDefined();
  });

  // 2. Image rendering test
  it('renders the icon image', () => {
    render(<ListItem content="Image test" />);
    const listItem = screen.getByText('Image test').closest('li');
    const image = listItem?.querySelector('img');
    expect(image).toHaveProperty(
      'src',
      'http://localhost:3000/images/icon-list.svg'
    );
  });

  // 3. Styles test
  it('applies the correct styles', () => {
    render(<ListItem content="Styles test" />);
    const listItem = screen.getByText('Styles test').closest('li')!;
    expect(listItem.className).toContain('grid');
    expect(listItem.className).toContain('grid-cols-[21px_1fr]');
    expect(listItem.className).toContain('gap-4');
    expect(listItem.className).toContain('w-full');
  });

  // 4. Accessibility test
  it('has no accessibility violations', async () => {
    const { container } = render(
      <ul>
        <ListItem content="Accessibility test" />
      </ul>
    );
    const result = await axe(container);
    expect(result).toHaveNoViolations();
  });

  // 5. Snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(<ListItem content="Snapshot test" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
