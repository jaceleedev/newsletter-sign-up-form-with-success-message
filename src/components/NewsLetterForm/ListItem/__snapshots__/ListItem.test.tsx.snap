// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListItem > matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="grid grid-cols-[21px_1fr] gap-4 w-full"
  >
    <img
      alt=""
      aria-hidden="true"
      class="mt-[2px]"
      data-nimg="1"
      decoding="async"
      fetchpriority="high"
      height="21"
      src="/images/icon-list.svg"
      style="color: transparent;"
      width="21"
    />
    <p
      class="text-body text-dark-navy"
    >
      Snapshot test
    </p>
  </li>
</DocumentFragment>
`;
